sys/mips/mips/trap.c

```diff
@@ -1455,7 +1455,7 @@ trapDump(char *msg)
  * Return the resulting PC as if the branch was executed.
  *
  * XXXRW: What about CHERI branch instructions?
- * XXXAR: This needs to be fixed for cjalr/cjr/ccall_fast
+ * XXXAR: This needs to be fixed for ccall_fast
```

Comment:
CCall no longer has a delay slot (I don't know if it technically still does but is required to be a NOP, but if so then any fault in the delay slot is fatal), so this can go too.

Reply:
I use this in fasttrap (not yet committed here) to determine the next address when a branch instruction is replaced with a breakpoint (so, the fault is not generated from the delay slot)

Reply:
So, there's a problem here. CJALR and CJR need to change PCC, not just PCC.cursor. MipsEmulateBranch currently just returns an address to shove into the current PCC. This works for the CBx instructions because the only capability aspect to them them is their condition, but their execution is just like a normal non-CHERI branch. CJALR and CJR need to actually install PCC. Moreover, CJALR needs to set the link register, which won't happen. I don't want MipsEmulateBranch to grow all this because it's not needed for its normal use case. You probably want your own fuller emulation. You also need to be very careful that your emulation actually executes the delay slot of any branches you have, because MipsEmulateBranch doesn't, that's up to the caller.

Reply:
Thanks for the clarifications.
However, is the address returned here in the case of CJALR/CJR correct?

Reply:
MipsEmulateBranch returns a trapf_pc_t, which should be a capability. So this should be fine since I believe the callers also use the capability value and not the address.

Reply:
Right.. the name retAddr threw me. We'll give confusing harder-to-debug faults (EPC pointing at the fetch, rather than the original CJALR etc) if userspace ever does something wrong, but that's already true for CBTS etc if you go out of bounds. But CJALR (and normal JALR/JAL/BAL) still won't be updating the link register here (and shouldn't), so this is the wrong function for fasttrap.

Reply:
I don't use this for emulation, but just to find the next address and place a breakpoint there. This is exactly the same thing ptrace does in  sys/mips/mips/pm_machdep.c:424. Is this still the wrong function?

Reply:
That should be fine then, if you then go on to actually execute the branch. This strikes me as something that breaks in a multi-threaded environment, though?

Reply:
I think my implementation of fasttrap will work also in a multithreaded environment (but not sure). We will discuss it in the fasttrap pr